community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/CountsStorageServiceImpl.java

```diff
@@ -0,0 +1,135 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.store.counts;
+
+import org.neo4j.kernel.impl.api.CountsVisitor;
+import org.neo4j.kernel.impl.store.counts.keys.CountsKeyFactory;
+import org.neo4j.kernel.impl.store.counts.keys.IndexSampleKey;
+import org.neo4j.kernel.impl.store.counts.keys.IndexStatisticsKey;
+import org.neo4j.kernel.impl.store.counts.keys.NodeKey;
+import org.neo4j.kernel.impl.store.counts.keys.RelationshipKey;
+import org.neo4j.kernel.internal.DatabaseHealth;
+import org.neo4j.register.Register;
+
+public class CountsStorageServiceImpl implements CountsStorageService
+{
+    private CountsStore countsStore;
+    private IndexStatsUpdater indexStatsUpdater;
+    private final UpdaterFactory updaterFactory = new UpdaterFactory();
+
+    @Override
+    public CountsSnapshot snapshot( long txId )
+    {
+        return countsStore.snapshot( txId );
+    }
+
+    @Override
+    public Updater updaterFor( long txId )
+    {
+        return updaterFactory.getUpdater( countsStore, txId );
+    }
+
+    public IndexStatsUpdater indexStatsUpdater()
+    {
+        return indexStatsUpdater;
+    }
+
+    @Override
+    public Updater apply( long txId )
+    {
+        return updaterFor( txId );
+    }
+
+    @Override
+    public Register.DoubleLongRegister nodeCount( int labelId, Register.DoubleLongRegister target )
+    {
+        long value = countsStore.get( CountsKeyFactory.nodeKey( labelId ) )[0];
+        target.write( 0, value );
+        return target;
+    }
+
+    @Override
+    public Register.DoubleLongRegister relationshipCount( int startLabelId, int typeId, int endLabelId,
+            Register.DoubleLongRegister target )
+    {
+        long value = countsStore.get( CountsKeyFactory.relationshipKey( startLabelId, typeId, endLabelId ) )[0];
+        target.write( 0, value );
+        return target;
+    }
+
+    @Override
+    public Register.DoubleLongRegister indexUpdatesAndSize( int labelId, int propertyKeyId,
+            Register.DoubleLongRegister target )
+    {
+        long[] values = countsStore.get( CountsKeyFactory.indexStatisticsKey( labelId, propertyKeyId ) );
+        target.write( values[0], values[1] );
+        return target;
+    }
+
+    @Override
+    public Register.DoubleLongRegister indexSample( int labelId, int propertyKeyId, Register.DoubleLongRegister target )
+    {
+        long[] values = countsStore.get( CountsKeyFactory.indexSampleKey( labelId, propertyKeyId ) );
+        target.write( values[0], values[1] );
+        return target;
+    }
+
+    @Override
+    public void accept( CountsVisitor visitor )
+    {
+        countsStore.forEach( ( countsKey, values ) -> {
```

Comment:
traversal is not atomic here, you can see/not see actual values of one atomic state